idp-binding-service/src/main/java/io/mosip/idp/binding/services/KeyBindingServiceImpl.java

```diff
@@ -0,0 +1,320 @@
+		final String transactionId = IdentityProviderUtil.createTransactionId(null);
+		BindingTransaction transaction = new BindingTransaction();
+		transaction.setIndividualId(bindingOtpRequest.getIndividualId());
+		transaction.setAuthTransactionId(IdentityProviderUtil.createTransactionId(null));
+		transaction.setAuthChallengeTypes(Arrays.asList("OTP"));
+		transaction = cacheUtilService.setTransaction(transactionId, transaction);
+
+		SendOtpResult sendOtpResult;
+		try {
+			sendOtpResult = keyBindingWrapper.sendBindingOtp(transaction.getAuthTransactionId(),
+					bindingOtpRequest.getIndividualId(), bindingOtpRequest.getOtpChannels(), headers);
+		} catch (SendOtpException e) {
+			log.error("Failed to send otp for transaction : {}", transactionId, e);
+			throw new IdPException(e.getErrorCode());
+		}
+
+		if (sendOtpResult == null || !transaction.getAuthTransactionId().equals(sendOtpResult.getTransactionId())) {
+			log.error("send-otp Failed with result : {}", sendOtpResult);
+			throw new IdPException(ErrorConstants.SEND_OTP_FAILED);
+		}
+
+		OtpResponse otpResponse = new OtpResponse();
+		otpResponse.setTransactionId(transactionId);
+		otpResponse.setMaskedEmail(sendOtpResult.getMaskedEmail());
+		otpResponse.setMaskedMobile(sendOtpResult.getMaskedMobile());
+		return otpResponse;
+	}
+
+	@Override
+	public WalletBindingResponse bindWallet(WalletBindingRequest walletBindingRequest, Map<String, String> headers) throws IdPException {
+		BindingTransaction bindingTransaction = cacheUtilService.getTransaction(walletBindingRequest.getTransactionId());
+		if (bindingTransaction == null)
+			throw new InvalidTransactionException();
+
+		if (!bindingTransaction.getIndividualId().equals(walletBindingRequest.getIndividualId()))
+			throw new IdPException(INVALID_INDIVIDUAL_ID);
+
+		if (walletBindingRequest.getChallengeList().stream().noneMatch(keyBindingAuthChallenge ->
+				bindingTransaction.getAuthChallengeTypes().contains(keyBindingAuthChallenge.getAuthFactorType())))
+			throw new IdPException(AUTH_FACTOR_MISMATCH);
+
+		String publicKey = IdentityProviderUtil.getJWKString(walletBindingRequest.getPublicKey());
+
+		KeyBindingResult keyBindingResult;
+		try {
+			keyBindingResult = keyBindingWrapper.doKeyBinding(bindingTransaction.getAuthTransactionId(),
+					walletBindingRequest.getIndividualId(), walletBindingRequest.getChallengeList(),
+					walletBindingRequest.getPublicKey(), headers);
+		} catch (KeyBindingException e) {
+			log.error("Failed to bind the key : {}", walletBindingRequest.getTransactionId(), e);
+			throw new IdPException(e.getErrorCode());
+		}
+
+		if (keyBindingResult == null || keyBindingResult.getCertificate() == null || keyBindingResult.getPartnerSpecificToken() == null) {
+			log.error("wallet binding failed with result : {}", keyBindingResult);
+			throw new IdPException(KEY_BINDING_FAILED);
+		}
+
+		PublicKeyRegistry publicKeyRegistry = storeKeyBindingDetailsInRegistry(walletBindingRequest.getIndividualId(),
```

Comment:
Will it make sense to move this to a different layer and call the layer from the the key binding wrapper ?  that way I feel we will be little consistent, since for validation it will be called from the auth wrapper.

Reply:
Or we have to make the call to validate binding also as internal not from the wrapper. Lets disuss this over call